src/server/zset_family.cc

```diff
@@ -1087,22 +1087,37 @@ void BZPopMinMax(CmdArgList args, Transaction* tx, SinkReplyBuilder* builder,
   return rb->SendNullArray();
 }
 
-vector<ScoredMap> OpFetch(EngineShard* shard, Transaction* t) {
+OpResult<vector<ScoredMap>> OpFetch(EngineShard* shard, Transaction* t, bool skip_dest_key) {
   ShardArgs keys = t->GetShardArgs(shard->shard_id());
   DCHECK(!keys.Empty());
 
+  ShardArgs::Iterator start = keys.begin(), end = keys.end();
+
+  if (skip_dest_key) {
+    // If destkey is only found on this shard we can skip it
+    if (++start == end) {
+      return OpStatus::SKIPPED;
+    }
+  }
+
   vector<ScoredMap> results;
-  results.reserve(keys.Size());
+  results.reserve(keys.Size() - (skip_dest_key ? 1 : 0));
 
   auto& db_slice = t->GetDbSlice(shard->shard_id());
-  for (string_view key : keys) {
-    auto it = db_slice.FindReadOnly(t->GetDbContext(), key, OBJ_ZSET);
-    if (!it) {
+  for (; start != end; ++start) {
+    auto it_res = db_slice.FindReadOnly(t->GetDbContext(), *start);
```

Comment:
FindReadOnly with object type will return wrong type status if the key is not from given type, so you can use this api just check the status and if not OK return status